### ISSUE TITLE
extract generated graphs

### ISSUE THREAD

**Issue by nahomtse** [mmiller96/netgan_pytorch#12]
Hi,

Do you know a way how to extract the new replicated graphs from this code and to export it?
So I want to have the graphs/adj. matrix generated with GAN?
I can not solve this problem.

Thank you in advance!

**Comment by mmiller96** [comment]
Hi,

I assume that the training worked. The class Trainer has a function create_graph(). 
The important part in this section is
""
num_iterations = int(num_samples/1000)+1
for j in range(num_iterations):
    if(j%10 == 1): print(j)
    samples.append(self.generator.sample_discrete(int(num_samples/1000), self.device))
samples = np.vstack(samples)
gr = utils.score_matrix_from_random_walks(samples, self.N)
gr = gr.tocsr()
_graph = utils.graph_from_scores(gr, self.graph.sum())
""
In short many random walks are generated, then the transition matrix gr is calculated (how many times an edge occured) and afterwards a graph is synthetically generated.
_graph is the synthetic graph, you could run "utils.graph_from_scores()" multiple times to get more synthetic graphs. 
With scipy.sparse.save_npz it should be possible to save the graph.

Best regards

**Comment by nahomtse** [comment]
Hi,

Thank you for your quick reply! 
I understand your code and know that _graph is the synthetic graph and gr is transition matrix. 
However, I can not call and run gr and _graph in the demo file (I'm not good with programming, I'm sorry). 
I hope you can add a small code (just to run 1 synthetic graph) to call these into the demo file or in this comment section. 

Best regards

**Comment by mmiller96** [comment]
Hi,

I am currently on vacation and I will be only back at 17.october. 
I guess there is a problem with the library versions but to be sure I have to dig deeper into the error messages.

Best regards

**Comment by nahomtse** [comment]
Hi,

Thanks again. 
I have appended a list to produce synthetic graphs, but could not figure out how to extract the graphs from the training file. 

Also, when using my own data and using different sizes of sparse graphs as input, I keep getting an error in the trainer line and the error is about the function used in utils, about split train/test/Val adjacency matrix. The error is about the line ' assert test_zeros.shape[0] == n_test' or the symmetrize coding lines. . I'm trying to figure out what these lines do and how to solve them, but I can not understand them properly. 

Hope you can help me out. 

Best regards,
Nahom Tsehaie

**Comment by mmiller96** [comment]
Hi,

I have simply taken the function "train_val_test_split_adjacency" from the original [github netgan tensorflow](https://github.com/danielzuegner/netgan) account. So I don't know in depth how the functions works. However, the problem seems to be easily solved when setting set_ops=False. It seems to be just an optimized version of getting the test_zeros values.

I have changed the trainer class so that set_ops is set False at default. 
Additionally I have created a new function "create_transitions_matrix" that simplifies the generation of graphs. 
I have added a new file "demo_full_example.ipynb" that shows a full work flow of generating synthetic graphs from an adjacency list which i have found [here](https://networkrepository.com/inf.php).

If there are still questions left, feel free to ask.


Best regards,
Markus Miller
